ovsdb: Use incremental change instead of overriding

cathay4t opened the pull request:
Previously, for OVS global config, nmstate is overriding the whole
`external_ids` and `other_config`, this patch changed to use `Mutate` 
operation to do incremental changes via `insert` and `delete` call.

Technical detail:

* OVS DB protocol is defined in RFC-7047.
* The global `ovs-db` section is handled by nmstate, not NM.
* The previous used `update` operation will override a matched row in
  table, meaning the whole `external_ids` or `other_config`. This patch
  only use `update` operation when purging.
* The `mutate` operation can insert and delete key-value without
  overriding the whole map.
* The `insert` mutator will not override if key already exist, so we
  use `delete` mutator first.

With this change, the complex merging action is not required in
`MergedOvsDbGlobalConfig` any more.

Existing OVS test cases are enough to test this change.

Resolves: https://issues.redhat.com/browse/RHEL-78652

cathay4t committed:
Run cargo clippy

Signed-off-by: Gris Ge <fge@redhat.com>

ihuguet commented on rust/src/lib/ovs.rs:
```diff
@@ -185,114 +185,19 @@ fn value_to_hash_map(
-            current.other_config.as_ref().unwrap_or(&empty_map).clone();
-
-        if let Some(desired) = &mut desired {
+        let mut desired = desired;
```
This copy is not needed, you can maintain the `mut` in the function argument and use it directly (I guess that this  is optimized by the compiler, but anyway...)

cathay4t replied:
Since the struct is moving into the function, require it to be mut seems overkill to me.
I see rust-netlink code are just taking object and make it as mut internally without asking caller to move a mut object to function.

ihuguet replied:
I don't care much about this because it was only to make the code a bit more compact. But in this case, the variable to be moved doesn't need to be `mut` in the caller. As it is being moved, it is guaranteed that the caller doesn't hold any other references to it. A non-mut variable can be moved to a mut variable directly. The variable needs to be mut in the caller when you pass a `&mut`, but not to move.

```
struct Test { a: u32}

fn main() {
    let t = Test {a: 10};
    print_test(t);
}

fn print_test(mut t: Test) {
    t.a = 20;
    println!("{}", t.a);
}
```

cathay4t replied:
You are right. My memory on this is wrong or outdated.

cathay4t replied:
Changed back to `mut desired`.

ihuguet commented on rust/src/lib/ovsdb/db.rs:
```diff
@@ -45,98 +33,84 @@ impl OvsDbCondition {
     pub(crate) fn new() -> Result<Self, NmstateError> {
         Ok(Self {
             rpc: OvsDbJsonRpc::connect(DEFAULT_OVS_DB_SOCKET_PATH)?,
+            transaction_id: get_sec_since_epoch(),
```
Using the time in seconds as unique ID doesn't seem a good idea. There might be 2 transactions within the same second. This transaction needs to be unique per connection to ovsdb? Or unique across all potential connections to ovsdb?
If it's the first, then it is OK, I guess, but why not to start with 0?
If not, what else can we use? A pseudo-random number?

ffmancera replied:
Good point, I checked the Open vSwitch RFC 7047 - https://www.rfc-editor.org/rfc/rfc7047#section-4.1

```
   The value of "id" MUST be unique among all in-flight transactions
   within the current JSON-RPC session.  Otherwise, the server may
   return a JSON-RPC error.
```

Therefore, we should not care about other potential connections to ovsdb.

cathay4t replied:
The ID only required to be unique inside a DB connection.
Changed to start with 0.

Good catch!

ihuguet commented on rust/src/lib/ovsdb/db.rs:
```diff
@@ -45,98 +33,84 @@ impl OvsDbCondition {
-                            );
-                        }
+
+        if let Value::Array(reply) = reply {
+            if let Some(entries) = reply
```
I think that this can be unified to `if let Value::Array(Some(entries)) = reply` (both `else` blocks are identical).

cathay4t replied:
They are not identical, the `reply` is different.

cathay4t replied:
It is `serde_json::Value(Vec<reply>)`, not `serde_json::Value(Option<Vec<reply>>)`.

cathay4t replied:
Never mind. I noticed `serde_json::Value::as_array()` can be use in chain actions.

cathay4t replied:
Changed to
```rust
        if let Some(entries) = reply
            .as_array()
            .and_then(|reply| reply.first())
            .and_then(|v| v.as_object())
            .and_then(|v| v.get("rows"))
            .and_then(|v| v.as_array())
        {
```

ihuguet replied:
Didn't look very carefully to this. Indeed, you cannot do what I said in this case. The modification is nice, though, thanks.

ihuguet commented on rust/src/lib/ovsdb/db.rs:
```diff
@@ -396,26 +315,29 @@ pub(crate) fn parse_uuid_array(v: &[Value]) -> Vec<String> {
+                    .as_object()
+                    .and_then(|r| r.get("details"))
+                    .and_then(|d| d.as_str())
+                    .unwrap_or("");
```
Maybe put here a generic error description instead of an empty string. "unknown error"?

ffmancera replied:
+1

cathay4t replied:
Changed to `.unwrap_or("unknown error");`. Thanks.

ihuguet commented on rust/src/lib/ovsdb/global_conf.rs:
```diff
@@ -33,46 +36,204 @@ fn convert_map(
+            })],
+        })?;
+
+        if let Value::Array(reply) = reply {
+            if let Some(global_conf) = reply
```
Same, this can be `if let Value::Array(Some(global_conf)) = reply`.

cathay4t replied:
Changed to
```rust
        if let Some(global_conf) = reply
            .as_array()
            .and_then(|reply| reply.first())
            .and_then(|v| v.as_object())
            .and_then(|v| v.get("rows"))
            .and_then(|v| v.as_array())
            .and_then(|v| v.first())
            .and_then(|v| v.as_object())
        {
```

ffmancera reviewed (approved):
LGTM, please check @ihuguet comments.

liangwen12year commented on rust/src/lib/ovsdb/global_conf.rs:
```diff
@@ -33,46 +36,193 @@ fn convert_map(
+    let delete_keys: Vec<Value> =
+        data.keys().map(|k| Value::String(k.to_string())).collect();
+    if !delete_keys.is_empty() {
+        mutations.push(OvsDbMutation {
```
Nit, can we add `log::debug!("Deleting keys from {}: {:?}", column, delete_keys);` here? If you think that it is unnecessary, please ignore this comment.

cathay4t replied:
Add `log::debug!("Removing old value of key {} from {}", k, column);`

liangwen12year commented on rust/src/lib/ovsdb/global_conf.rs:
```diff
@@ -33,46 +36,193 @@ fn convert_map(
+        .iter()
+        .filter_map(|(k, v)| {
+            v.as_ref().map(|v| {
+                Value::Array(vec![
```
How about also logging the keys being inserted?

cathay4t replied:
Added `log::debug!("Inserting key {} value {} into {}", k, v, column);`

liangwen12year commented on rust/src/lib/ovsdb/global_conf.rs:
```diff
@@ -33,46 +36,193 @@ fn convert_map(
+
+    let mut cli = OvsDbConnection::new()?;
+    let mut operations = Vec::new();
+    let mut is_extenal_ids_purged = false;
```
typo in the variable here.

liangwen12year replied:
should be `is_external_ids_purged`?

cathay4t commented:
New patch uploaded. Changes:
 * Add debug for purging, inserting, removing OVS entries.
 * Fixed the variable typo of `is_extenal_ids_purged`

maiqueb commented on rust/src/lib/ovsdb/global_conf.rs:
```diff
@@ -33,46 +36,200 @@ fn convert_map(
+
+    let mut cli = OvsDbConnection::new()?;
+    let mut operations = Vec::new();
+    let mut is_external_ids_purged = false;
```
you use this variable to control if you purge *both* the `external_ids` **and** the `other_config` tables.

Could we name this something a bit more generic ? e.g. `is_global_table_purged` or something like that ? 

This is a nit you can ignore.

cathay4t replied:
No. Only `external_ids` purged require OVN reinstall. The `is_external_ids_purged` is true when full purging or `external_ids` is set to purging.

maiqueb reviewed (approved):
Keep in mind I haven't been writing code in rust for some time ...

But LGTM.

cathay4t committed:
ovsdb: Use incremental change instead of overriding

Previously, for OVS global config, nmstate is overriding the whole
`external_ids` and `other_config`, this patch changed to use `Mutate`
operation to do incremental changes via `insert` and `delete` call.

Technical detail:

 * OVS DB protocol is defined in RFC-7047.
 * The global `ovs-db` section is handled by nmstate, not NM.
 * The previous used `update` operation will override a matched row in
   table, meaning the whole `external_ids` or `other_config`. This patch
   only use `update` operation when purging.
 * The `mutate` operation can insert and delete key-value without
   overriding the whole map.
 * The `insert` mutator will not override if key already exist, so we
   use `delete` mutator first.

With this change, the complex merging action is not required in
`MergedOvsDbGlobalConfig` any more.

Existing OVS test cases are enough to test this change.

Resolves: https://issues.redhat.com/browse/RHEL-78652

Signed-off-by: Gris Ge <fge@redhat.com>
